service/history/workflow/mutable_state_impl_restart_activity_test.go

```diff
@@ -0,0 +1,395 @@
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	BackoffIntervalCalculator = func(
+		time.Time,
+		int32,
+		int32,
+		*durationpb.Duration,
+		*durationpb.Duration,
+		*timestamppb.Timestamp,
+		float64,
+		*failurepb.Failure,
+		[]string,
+	) (time.Duration, enumspb.RetryState)
+
+	snapshot struct {
+		mutableStateApproximateSize int
+		activitySize                int
+	}
+
+	retrySuite struct {
+		suite.Suite
+
+		controller       *gomock.Controller
+		mockConfig       *configs.Config
+		mockShard        *shard.ContextTest
+		mockEventsCache  *events.MockCache
+		onActivityCreate *snapshot
+
+		mutableState               *MutableStateImpl
+		logger                     log.Logger
+		testScope                  tally.TestScope
+		originalGetBackoffInterval BackoffIntervalCalculator
+		backoffCalculatorStub      getBackoffIntervalStub
+		activity                   *persistencespb.ActivityInfo
+		failure                    *failurepb.Failure
+		timeSource                 TimeSourceStub
+	}
+)
+
+func TestMutableStateRetryActivitySuite(t *testing.T) {
+	s := new(retrySuite)
+
+	suite.Run(t, s)
+}
+
+func (s *retrySuite) SetupSuite() {
+	s.mockConfig = tests.NewDynamicConfig()
+	// set the checksum probabilities to 100% for exercising during test
+	s.mockConfig.MutableStateChecksumGenProbability = func(namespace string) int { return 100 }
+	s.mockConfig.MutableStateChecksumVerifyProbability = func(namespace string) int { return 100 }
+	s.mockConfig.MutableStateActivityFailureSizeLimitWarn = func(namespace string) int { return 1 * 1024 }
+	s.mockConfig.MutableStateActivityFailureSizeLimitError = func(namespace string) int { return 2 * 1024 }
+
+	s.originalGetBackoffInterval = getBackoffInterval
+	getBackoffInterval = s.backoffCalculatorStub.getBackoffInterval
+}
+
+func (s *retrySuite) TearDownSuite() {
+	getBackoffInterval = s.originalGetBackoffInterval
+}
+
+func (s *retrySuite) SetupTest() {
+
+	s.controller = gomock.NewController(s.T())
+	s.mockEventsCache = events.NewMockCache(s.controller)
+
+	s.mockShard = shard.NewTestContext(
+		s.controller,
+		&persistencespb.ShardInfo{
+			ShardId: 0,
+			RangeId: 1,
+		},
+		s.mockConfig,
+	)
+	s.mockShard.SetEventsCacheForTesting(s.mockEventsCache)
+
+	s.testScope = s.mockShard.Resource.MetricsScope.(tally.TestScope)
+	s.logger = s.mockShard.GetLogger()
+
+	s.mutableState = NewMutableState(s.mockShard, s.mockEventsCache, s.logger, tests.LocalNamespaceEntry, time.Now().UTC())
+	s.activity = s.makeActivityAndPutIntoFailingState()
+	s.failure = s.activityFailure()
+	s.backoffCalculatorStub.onNextCallExpect(
+		s.timeSource.Now(),
+		s.activity.Attempt,
+		s.activity.RetryMaximumAttempts,
+		s.activity.RetryInitialInterval,
+		s.activity.RetryMaximumInterval,
+		s.activity.RetryExpirationTime,
+		s.activity.RetryBackoffCoefficient,
+		s.failure,
+		s.activity.RetryNonRetryableErrorTypes,
+	)
+}
+
+func (s *retrySuite) TearDownTest() {
+	s.controller.Finish()
```

Comment:
You don't need this line.

Reply:
Why do you think I do not need it? I have expectations set during tests. Shouldn't I call `Finish` to make sure that all the expected calls actually happened?

Reply:
You do need this for gomock, and you can see it in other places throughout our codebase @alexshtin

Reply:
Oh nice! We can remove this from our other tests then